src/encoded/upgrade/genetic_modification.py

```diff
@@ -185,4 +185,78 @@ def genetic_modification_5_6(value, system):
     # https://encodedcc.atlassian.net/browse/ENCD-4042
     if value['purpose'] == 'validation':
-        value['purpose'] = 'characterization'
\ No newline at end of file
+        value['purpose'] = 'characterization'
+
+
+@upgrade_step('genetic_modification', '7', '8')
+def genetic_modification_7_8(value, system):
+    # https://encodedcc.atlassian.net/browse/ENCD-4349
+    import re
+    import copy
+    import time
+    import json
+
+    conn = system['registry'][CONNECTION]
+
+    sources = {
+        'addgene': r'^\d{5,6}$',
+        'bacpac': r'^([A-Z]{2,3}\d{2,3}|[A-Z]{3})-\d{1,4}[A-Z]\d{1,2}$',
+        'dharmacon': r'^[DL]-\d{6}-\d{2}(-\d{2,4})?$',
+        'hugo-bellen': r'^MI\d{5}$',
+        'human-orfeome': r'^([A-Z]{2})?\d{1,9}$',
+        'plasmid-repository': r'^HsCD\d{8}$',
+        'sigma': r'^[A-Z]{3}\d{3}$',
+        'source-bioscience': r'^[A-Z]{3}\d{3,4}[a-z][A-Z]\d{2}(_[A-Z]\d{2})?$',
+        'thermo-fisher': r'^[a-zA-Z]{1,3}\d{5,6}$',
+        'trc': r'^TRCN\d{10}$',
+        'brenton-graveley': r'^BGC#\d{7}$'
+    }
+    reagents = value.get('reagents')
+    if reagents:
+        new_reagents = []
+        unmatched_reagents = ''
+        for reagent in reagents:
+            identifier = reagent['identifier']
+            source_id = reagent['source']
+            source_obj = conn.get_by_uuid(source_id)
+            source_name = source_obj.properties['name']
+            new_reagent = copy.deepcopy(reagent)
+            matching = []
+            source_from_reagent = None
+            for source in sources:
+                if source in source_name:
+                    source_from_reagent = source_name
+                    break
+            # Preferentially add prefix to reagent that matches the reagent source
+            if source_from_reagent is not None:
+                regex = sources[source_from_reagent]
+                if re.match(regex, identifier) is not None:
+                    matching.append((source_name, source_from_reagent, identifier))
+            else:
+                for source, regex in sources.items():
+                    if re.match(regex, identifier) is not None:
+                        matching.append((source_name, source, identifier))
+            if not matching:
+                unmatched_reagents += '{} '.format(json.dumps(reagent))
+            else:
+                if len(matching) >= 2:
+                    # Bring the identifier with matching source to the reagent to the top
+                    matching.sort(key=lambda match: int(match[1] in match[0]), reverse=True)
+                    # matching = [i for i in matching if i[1] in i[0]]
+                _, source, identifier = matching[0]
+                new_identifier = ('{}:{}'.format(source, identifier))
+                new_reagent['identifier'] = new_identifier
+                new_reagents.append(new_reagent)
+        # Dump the unmatched identifier reagents to the notes
+        if unmatched_reagents:
```

Comment:
This block of code is different, it creates value['notes'] based on unmatched reagents.  But it sitting in between code that acts on value['reagents'].  But they are not dependent on each other.  I might finish with the value reagents and then create the value notes at the end.

Reply:
Rearranged